### PR TITLE
125 nan gradients from pfaffian computation when determinant is zero x025 singularity

### LINKED ISSUE #125: NaN gradients from pfaffian computation when determinant is zero (x^0.25 singularity)
## Summary

 When computing transition amplitudes via `|det(O)|^{1/4}` (fourth root of absolute determinant), zero or near-zero determinants produce **NaN gradients** that corrupt the entire backward pass. This affects any use case
  that takes a fractional power of the pfaffian/determinant for gradient-based optimization.

 ## Background

 In our downstream project [FermiDL](https://github.com/Mamarco-Armenta/FermiDL), we build weight matrices from non-interacting fermionic circuits using MatchCake. The amplitude `|<l|U(θ)|j>|` is computed as
 `|det(O)|^{1/4}` (since `Pf(O) = sqrt(|det(O)|)` and amplitude = `sqrt(Pf(O))`).

 Parity-forbidden transitions (where `popcount(l) ≠ popcount(j) mod 2`) produce determinants that are numerically zero or near-zero (~1e-32 residuals).

 ## The Problem

 The derivative of `f(x) = x^{0.25}` is:

 f'(x) = 0.25 * x^{-0.75} = 0.25 / x^{0.75}

 At `x = 0`, this is **infinite**. In PyTorch:

 1. Forward pass: `0^{0.25} = 0` ✓ (correct)
 2. Backward pass: `0.25 * 0^{-0.75}` = **NaN** ✗

 Even when these entries are later multiplied by zero (e.g., via a parity mask), the NaN propagates because `NaN * 0 = NaN` in IEEE 754 arithmetic. This corrupts all parameter gradients.

 ## Current State in MatchCake

 MatchCake's pfaffian functions use an additive epsilon approach:

 **`matchcake/utils/_pfaffian.py`** (lines ~300, 305, 333):
 ```python
 pf = sqrt(|det| + epsilon)  # epsilon=1e-12
 matchcake/utils/torch_pfaffian.py (line ~11):
 pf = torch.sqrt(torch.abs(det) + Pfaffian.EPSILON)  # EPSILON=1e-10
```
 This prevents NaN for sqrt (exponent 0.5), but does not protect downstream consumers who may take further fractional powers of the pfaffian (e.g., pf^{0.5} to get amplitudes, which is equivalent to |det|^{0.25}).

 Our Workaround.

 In FermiDL, we changed:

 # BEFORE (NaN gradients):
```python
amplitudes = torch.pow(torch.clamp(torch.abs(dets), min=0.0), 0.25)
```

 # AFTER (fixed):
```python
 amplitudes = torch.pow(torch.clamp(torch.abs(dets), min=1e-32), 0.25)
```

 The clamp floor of 1e-32 ensures gradients are large but finite (~10^{24}), and when multiplied by a zero parity mask, they cleanly produce 0.

 Suggested Fix

 Consider one or more of:

 1. Document the singularity: Warn users that fractional powers of pfaffians/determinants require an epsilon floor to avoid NaN gradients at zero.
 2. Add a clamp_min parameter to pfaffian functions (default 1e-32 or similar) so consumers don't have to handle this themselves.
 3. Provide a numerically safe amplitude function that computes |det|^{1/4} (or arbitrary |det|^{p}) with built-in gradient safety.
 4. Use torch.clamp instead of additive epsilon in existing pfaffian functions — sqrt(clamp(|det|, min=eps)) is mathematically cleaner than sqrt(|det| + eps) because it doesn't bias non-zero values.

 Environment

 - MatchCake version: 0.2.2 (dev branch, commit 1558e55)
 - PyTorch: >=2.6
 - Python: 3.12

### PR DESCRIPTION
**Pull request by JeremieGince**
# Description

This pull request improves the numerical stability and robustness of the Pfaffian computation utilities and adds new tests for edge cases. The main changes include reducing the default epsilon for numerical clipping, updating the determinant-to-Pfaffian computation to use clipping instead of addition, and expanding the test suite to cover zero matrices and their gradients.

**Numerical stability improvements:**

* Reduced the default `epsilon` parameter from `1e-12` to `1e-32` in `pfaffian`, `pfaffian_by_det`, and `pfaffian_by_det_cuda` to allow for more precise numerical clipping in the computation of the Pfaffian. [[1]](diffhunk://#diff-a35727cb8592a9a8f9010ce4ef41daae713b016f05023fe86cec2ea6e71f3899L18-R37) [[2]](diffhunk://#diff-a35727cb8592a9a8f9010ce4ef41daae713b016f05023fe86cec2ea6e71f3899L50-R68) [[3]](diffhunk://#diff-a35727cb8592a9a8f9010ce4ef41daae713b016f05023fe86cec2ea6e71f3899L75-R99)
* Changed the determinant-to-Pfaffian calculation to use `qml.math.clip`/`qml.math.clamp` for lower bounding the determinant's absolute value, instead of adding epsilon, which improves numerical robustness for very small determinants. [[1]](diffhunk://#diff-a35727cb8592a9a8f9010ce4ef41daae713b016f05023fe86cec2ea6e71f3899L36-R57) [[2]](diffhunk://#diff-a35727cb8592a9a8f9010ce4ef41daae713b016f05023fe86cec2ea6e71f3899L65-R89)
* Ensured that the epsilon value is also correctly set for the `PfaffianFDBPf` method by assigning it to `torch_pfaffian.PfaffianStrategy.EPSILON`.

**Documentation:**

* Added a detailed docstring to `pfaffian_by_det` explaining its parameters, return value, and behavior for different matrix types and backends.

**Testing improvements:**

* Added tests for computing the Pfaffian of zero matrices and checking that their squared Pfaffian is zero, as well as corresponding gradient checks for these cases.
* Moved `torch` and `gradcheck` imports to the top of the test file for consistency. [[1]](diffhunk://#diff-218989875e2fa3e92cc4b677be41884bdb3446825b7637efaa90ef4dbf685e27R5-R6) [[2]](diffhunk://#diff-218989875e2fa3e92cc4b677be41884bdb3446825b7637efaa90ef4dbf685e27L45-L47)

Related Issue: https://github.com/MatchCake/MatchCake/issues/125

------------------------------------------------------------------------------------------------------------

# Checklist

Please complete the following checklist when submitting a PR. The PR will not be reviewed until all items are checked.

- [x] All new features include a unit test.
      Make sure that the tests passed and the coverage is
      sufficient by running 
      `uv run pytest tests --cov=src --cov-report=term-missing --durations=30 --session-timeout=600`.
- [x] All new functions and code are clearly documented.
- [x] The code is formatted using Black.
      You can do this by running `uv run black src tests`.
- [x] The imports are sorted using isort.
      You can do this by running `uv run isort src tests`.
- [x] The code is type-checked using Mypy.
      You can do this by running `uv run mypy src tests`.

### CODE REVIEW

**Review by MarcoArmenta — CHANGES_REQUESTED** [review]
Thanks @JeremieGince! I ran PR #128 locally against the FermiDL/uv environment (Python 3.12, torch 2.10, pennylane 0.44.x, torchpfaffian 0.0.0) and verified:

- All 25 tests in `tests/test_utils/test_pfaffian.py` pass, including the new `test_with_zeros` and `test_grads_with_zeros` for both `method="det"` and `method="PfaffianFDBPf"`.
- The downstream-fractional-power scenario from issue #125 (`utils.pfaffian(zero_matrix) ** 0.5`) is now NaN-free with PR #128 — `pf = 1e-16`, `pf**0.5 = 1e-8`, and the gradient w.r.t. the input matrix is finite (in fact zero, because clamp's gradient is zero at the boundary, which kills the chain back to the matrix).

**However, I should flag something:** PR #128 does not actually exercise the code path that triggered the NaN in my FermiDL workflow. FermiDL computes `torch.linalg.det(...)` directly and never calls `matchcake.utils.pfaffian` (see `src/fermidl/modules/nif_linear.py:226-227` in the FermiDL repo). The local fix in FermiDL (`torch.pow(torch.clamp(torch.abs(dets), min=1e-32), 0.25)`) is what unblocked me. PR #128 is still a real improvement — it protects any consumer that does call `utils.pfaffian` and then takes further fractional powers — but it doesn't patch my code path, so we should be clear that issue #125's suggestion #1 (documentation) and suggestion #3 (a safe `|det|^p` amplitude helper) remain open.

**Two small requests before merge:**

1. **Docstring**: The `pfaffian_by_det` docstring describes the inner clip but doesn't warn downstream consumers about further fractional powers. Could you add a short note like:

   > Note: this protects the inner `sqrt(|det|)` gradient. If you take further fractional powers of the return value (e.g., `pf**0.5` to recover `|det|^{1/4}` amplitudes), apply an additional clamp at the call site, since the inner clip alone is not sufficient when the consumer's gradient path bypasses this function.

2. **Thread-safety**: `torch_pfaffian.PfaffianStrategy.EPSILON = epsilon` at line 128 is a global write executed on every call. Concurrent or nested `pfaffian(..., method="PfaffianFDBPf")` calls with different epsilon values would race. If `torch_pfaffian.get_pfaffian_function` accepts an `epsilon` kwarg, pass it through instead; otherwise add a thread-local/context-manager wrapper, or at minimum a comment flagging the assumption that this is called from a single thread.

**Heads up — silent behavior change:** the formula switch from `sqrt(|det| + 1e-12)` to `sqrt(clamp(|det|, min=1e-32))` is a bias reduction (good) but pf values for small-but-nonzero dets differ by up to ~1e-6 between old and new code. Worth a one-line note in the PR description so downstream users who pin against numerical regressions know to expect it.

After those two changes I'm happy to approve. I'll leave issue #125 open after merge with a note about the two remaining unaddressed suggestions (docs warning + optional safe-amplitude helper), or open a follow-up issue, whichever you prefer.

Cheers

**Comment by JeremieGince** [comment]
Thanks for the review @MarcoArmenta!

**On the docstring:** I'm not sure I see the value of that note in this context. The docstring for `pfaffian_by_det` describes what the function does. Warning about downstream usage patterns that are entirely outside its scope feels out of place. If a consumer takes further fractional powers of the return value, that's their responsibility to clamp. I'd rather keep the docstring focused on the function's own behavior.

**On thread-safety:** I've wrapped the global write in a context manager as a hotfix. That said, this is really something that should be handled properly inside `TorchPfaffian` itself in the future. More importantly, MatchCake does not use `PfaffianFDBPf` by default. The default method is `det`, so this is a non-issue for the vast majority of users.

Let me know if that works for you.
